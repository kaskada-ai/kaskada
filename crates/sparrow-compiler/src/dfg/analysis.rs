//! Implement the DFG analyses for the E-graph library.
//!
//! [`DfgAnalysis`] implements the [`Analysis`] trait which instructs
//! the E-graph library how to create [`DfgAnalysisData`] for each node
//! added to the DFG and merge information between nodes when they are
//! determined to be equivalent.
use egg::{Analysis, Id, Language};
use hashbrown::HashMap;
use itertools::Itertools;
use smallvec::smallvec;
use sparrow_arrow::scalar_value::ScalarValue;
use sparrow_instructions::{InstKind, InstOp};

use super::{DfgGraph, DfgLang, StepKind};
use crate::dfg::Expression;

/// The E-graph analysis used to create and merge [`DfgAnalysisData`].
#[derive(Default, Debug)]
pub(crate) struct DfgAnalysis;

/// Analysis data that is added to each DFG node.
///
/// This is computed from the DFG nodes, and propagated along the nodes
/// as equivalent nodes are discovered, etc.
///
/// Some information (such as shape and type) is a little less intuitive
/// here than if it were specified while constructing the node. However,
/// having to specify it when creating the node would make simplification
/// more difficult, since every simplification rule would need to specify
/// the new shape/type information.
#[derive(Debug, Clone, PartialEq)]
pub(crate) struct DfgAnalysisData {
    /// The node defining the operation (domain) this node is in.
    domain: Domain,
    /// The literal value equivalent to this node, if any.
    literal: Option<ScalarValue>,
    /// If this node has known fields.
    known_fields: Option<HashMap<Id, Id>>,
    /// The type of node
    kind: StepKind,
    /// Children to the node
    children: Vec<Id>,
}

#[derive(Debug, Clone, PartialEq)]
enum Domain {
    /// The DFG node is an operation, and is thus its own operation.
    /// This needs to be a separate analysis, since the node hasn't been
    /// created (assigned an ID) yet.
    Operation,
    /// The DFG node is an expression associated with the given operation ID.
    Expression(Id),
}

impl DfgAnalysisData {
    pub(crate) fn literal_opt(&self) -> Option<&ScalarValue> {
        self.literal.as_ref()
    }

    /// Return the operation of this node.
    ///
    /// It must be passed it's own ID since that wasn't available when
    /// the node was created, and for operation nodes, the ID of the
    /// operation is the ID of the node.
    pub(crate) fn operation(&self, self_id: Id) -> Id {
        match self.domain {
            Domain::Operation => self_id,
            Domain::Expression(id) => id,
        }
    }

    pub(crate) fn kind(&self) -> StepKind {
        self.kind.clone()
    }
}

impl Analysis<DfgLang> for DfgAnalysis {
    type Data = DfgAnalysisData;

    fn merge(&mut self, a: &mut Self::Data, b: Self::Data) -> egg::DidMerge {
        let mut a_modified = false;
        let mut b_modified = false;

        // Merge literal
        match (a.literal_opt(), b.literal_opt()) {
            (None, Some(literal)) => {
                a.literal = Some(literal.clone());
                a_modified = true;
            }
            (Some(_), None) => {
                b_modified = true;
            }
            (None, None) => {
                // Nothing to change
            }
            (Some(lit_a), Some(lit_b)) if lit_a.is_null() && lit_b.is_null() => {
                // This is weird -- they may be *different scalars* but both
                // null. In theory, we shouldn't be merging them, but it could
                // happen.
            }
            (Some(lit_a), Some(lit_b)) => {
                assert_eq!(lit_a, lit_b, "Literals must be equal when merging.")
            }
        }

        egg::DidMerge(a_modified, b_modified)
    }

    fn make(egraph: &DfgGraph, enode: &DfgLang) -> Self::Data {
        // If this is a record creation, we have known field values.
        let known_fields =
            if let StepKind::Expression(Expression::Inst(InstKind::Record)) = enode.kind() {
                Some(enode.children().iter().copied().tuples().collect())
            } else {
                None
            };

        let literal = match enode.kind() {
            StepKind::Expression(Expression::Literal(value)) => Some(value.clone()),
            // If all of the arguments to an instruction (omitting the operation) are constant,
            // then the expression can be constant evaluated.
            StepKind::Expression(Expression::Inst(inst))
                if enode.children()[0..enode.children().len() - 1]
                    .iter()
                    .all(|arg| egraph[*arg].data.literal_opt().is_some()) =>
            {
                // If all the inputs are literals, just evaluate the node to get a literal.
                let literals: Vec<_> = enode.children()[0..enode.children().len() - 1]
                    .iter()
                    .map(|id| egraph[*id].data.literal_opt().unwrap().clone())
                    .collect();
                let literal = super::const_eval::evaluate_constant(inst, literals)
                    .expect("Result of constant evaluation");
                Some(literal)
            }
            StepKind::Transform => egraph[enode.children()[0]].data.literal_opt().cloned(),
            _ => None,
        };

        let domain = match enode.kind() {
            StepKind::Operation(_) => Domain::Operation,
            StepKind::Expression(_) | StepKind::Transform => {
                let operation = enode.children().last().unwrap_or_else(|| {
                    panic!(
                        "expressions must have at least one argument but {:?} did not",
                        enode.kind(),
                    )
                });
                Domain::Expression(*operation)
            }
            StepKind::Error => Domain::Operation,
            StepKind::Window(_) => Domain::Operation,
        };

        let children: Vec<Id> = enode.children().to_vec();

        DfgAnalysisData {
            domain,
            literal,
            known_fields,
            kind: enode.kind().clone(),
            children,
        }
    }

    fn modify(egraph: &mut DfgGraph, id: Id) {
        let data = &egraph[id].data;

        let operation = data.operation(id);
        let mut to_merge = Vec::new();

        let is_literal = if let Some(value) = data.literal_opt().cloned() {
            to_merge.push(EquivalentTo::Literal(value));
            true
        } else {
            false
        };

        for node in egraph[id].nodes.iter() {
            if !is_literal
                && matches!(node.kind(), StepKind::Expression(Expression::Inst(_)))
                && node
                    .children()
                    .iter()
                    .all(|arg| egraph[*arg].data.literal_opt().is_some())
            {
                if let StepKind::Expression(Expression::Inst(inst)) = node.kind() {
                    // If all the inputs are literals, just evaluate the node to get a literal.
                    let literals: Vec<_> = node
                        .children()
                        .iter()
                        .map(|id| egraph[*id].data.literal_opt().unwrap().clone())
                        .collect();
                    let literal = super::const_eval::evaluate_constant(inst, literals)
                        .expect("Result of constant evaluation");
                    to_merge.push(EquivalentTo::Literal(literal));
                    break;
                }
            } else if let StepKind::Expression(Expression::Inst(InstKind::FieldRef)) = node.kind() {
                // For each first field ref with `known_fields` equivalent to this `id`,
                // Union this `id` and the corresponding known field.

                let base = node.children()[0];
                let field = node.children()[1];

                // If the base has known fields, we can dereference them.
                if let Some(base_known_fields) = &egraph[base].data.known_fields {
                    let field_value = base_known_fields.get(&field).unwrap_or_else(|| {
                        panic!("Missing field value for {field} in {base_known_fields:?}")
                    });
                    to_merge.push(EquivalentTo::Node(*field_value));
                }

                // If the base is a `null_if` or an `if` we can push the field-ref inside.
                //
                // - `null_if(cond, base).field = null_if(cond, base.field)`
                // - `if(cond, base).field = if(cond, base.field)`.
                //
                // We can't (currently) express these easily using field refs since we can't
                // match on the field name.
                for base_node in egraph[base].nodes.iter() {
                    match base_node.kind() {
                        StepKind::Expression(Expression::Inst(InstKind::Simple(InstOp::If))) => {
                            to_merge.push(EquivalentTo::If {
                                cond: base_node.children()[0],
                                base: base_node.children()[1],
                                field,
                            });
                        }
                        StepKind::Expression(Expression::Inst(InstKind::Simple(
                            InstOp::NullIf,
                        ))) => {
                            to_merge.push(EquivalentTo::NullIf {
                                cond: base_node.children()[0],
                                base: base_node.children()[1],
                                field,
                            });
                        }
                        _ => (),
                    }
                }
            }
        }

        for equivalent in to_merge {
            match equivalent {
                EquivalentTo::Literal(literal) => {
                    let literal = egraph.add(DfgLang::new(
                        StepKind::Expression(Expression::Literal(literal)),
                        smallvec![operation],
                    ));
                    egraph.union(id, literal);
                }
                EquivalentTo::Node(equivalent) => {
                    egraph.union(id, equivalent);
                }
                EquivalentTo::NullIf { cond, base, field } => {
                    let cond = into_operation(egraph, cond, operation);
                    let base = into_operation(egraph, base, operation);
                    let field = into_operation(egraph, field, operation);
                    let base_field = egraph.add(DfgLang::new(
                        StepKind::Expression(Expression::Inst(InstKind::FieldRef)),
                        smallvec![base, field, operation],
                    ));
                    let equivalent = egraph.add(DfgLang::new(
                        StepKind::Expression(Expression::Inst(InstKind::Simple(InstOp::NullIf))),
                        smallvec![cond, base_field, operation],
                    ));
                    egraph.union(id, equivalent);
                }
                EquivalentTo::If { cond, base, field } => {
                    let cond = into_operation(egraph, cond, operation);
                    let base = into_operation(egraph, base, operation);
                    let field = into_operation(egraph, field, operation);

                    let base_field = egraph.add(DfgLang::new(
                        StepKind::Expression(Expression::Inst(InstKind::FieldRef)),
                        smallvec![base, field, operation],
                    ));
                    let equivalent = egraph.add(DfgLang::new(
                        StepKind::Expression(Expression::Inst(InstKind::Simple(InstOp::If))),
                        smallvec![cond, base_field, operation],
                    ));
                    egraph.union(id, equivalent);
                }
            }
        }
    }
}

fn into_operation(egraph: &mut DfgGraph, desired: Id, operation: Id) -> Id {
    let node = &egraph[desired];
    if node.data.operation(desired) == operation {
        desired
    } else if let Some(literal) = node.data.literal_opt().cloned() {
        egraph.add(DfgLang::new(
            StepKind::Expression(Expression::Literal(literal)),
            smallvec![operation],
        ))
    } else {
        egraph.add(DfgLang::new(
            StepKind::Transform,
            smallvec![desired, operation],
        ))
    }
}

enum EquivalentTo {
    Literal(ScalarValue),
    Node(Id),
    NullIf { cond: Id, base: Id, field: Id },
    If { cond: Id, base: Id, field: Id },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn merge_same() {
        // Merging is used when two nodes are identified as equal. We test it directly
        // here.
        let mut a = DfgAnalysisData {
            domain: Domain::Operation,
            literal: None,
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };
        let b = DfgAnalysisData {
            domain: Domain::Operation,
            literal: None,
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };

        let mut analysis = DfgAnalysis;
        let merge_result = analysis.merge(&mut a, b);
        assert!(!merge_result.0);
        assert!(!merge_result.1);
        assert_eq!(
            a,
            DfgAnalysisData {
                domain: Domain::Operation,
                literal: None,
                known_fields: None,
                kind: StepKind::Error,
                children: Vec::new(),
            }
        );
    }

    #[test]
    fn merge_lhs_literal() {
        // Merging is used when two nodes are identified as equal. We test it directly
        // here.
        let mut a = DfgAnalysisData {
            domain: Domain::Operation,
            literal: Some(ScalarValue::Int64(Some(5))),
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };
        let b = DfgAnalysisData {
            domain: Domain::Operation,
            literal: None,
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };

        let mut analysis = DfgAnalysis;
        let merge_result = analysis.merge(&mut a, b);
        assert!(!merge_result.0);
        assert!(merge_result.1);
        assert_eq!(
            a,
            DfgAnalysisData {
                domain: Domain::Operation,
                literal: Some(ScalarValue::Int64(Some(5))),
                known_fields: None,
                kind: StepKind::Error,
                children: Vec::new(),
            }
        );
    }

    #[test]
    fn merge_rhs_literal() {
        // Merging is used when two nodes are identified as equal. We test it directly
        // here.
        let mut a = DfgAnalysisData {
            domain: Domain::Operation,
            literal: None,
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };
        let b = DfgAnalysisData {
            domain: Domain::Operation,
            literal: Some(ScalarValue::Int64(Some(5))),
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };

        let mut analysis = DfgAnalysis;
        let merge_result = analysis.merge(&mut a, b);
        assert!(merge_result.0);
        assert!(!merge_result.1);
        assert_eq!(
            a,
            DfgAnalysisData {
                domain: Domain::Operation,
                literal: Some(ScalarValue::Int64(Some(5))),
                known_fields: None,
                kind: StepKind::Error,
                children: Vec::new(),
            }
        );
    }

    #[test]
    #[should_panic(expected = "Literals must be equal when merging")]
    pub fn test_merge_incompatible_literal() {
        // Merging is used when two nodes are identified as equal. We test it directly
        // here.
        let mut a = DfgAnalysisData {
            domain: Domain::Operation,
            literal: Some(ScalarValue::Int64(Some(5))),
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };
        let b = DfgAnalysisData {
            domain: Domain::Operation,
            literal: Some(ScalarValue::Int64(Some(6))),
            known_fields: None,
            kind: StepKind::Error,
            children: Vec::new(),
        };

        let mut analysis = DfgAnalysis;
        let merge_result = analysis.merge(&mut a, b);
        assert!(merge_result.0);
        assert!(!merge_result.1);
    }
}
