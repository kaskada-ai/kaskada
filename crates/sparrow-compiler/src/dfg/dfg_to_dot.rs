//! Utilities for printing a DFG to DOT (graphviz).

use std::borrow::Cow;

use itertools::Itertools;
use sparrow_instructions::InstKind;

use super::StepKind;
use crate::dfg::{DfgExpr, Expression};

/// Write the DFG for the given expression to `w` in DOT format.
///
/// This uses a subgraph for each operation.
pub(super) fn dfg_expr_to_dot(expr: &DfgExpr, w: &mut impl std::io::Write) -> anyhow::Result<()> {
    writeln!(w, "digraph dfg {{")?;
    writeln!(w, "rankdir = \"BT\";  ")?;

    for (operation, ids) in &expr.ids().group_by(|id| expr.operation(*id)) {
        if let Some(operation) = operation {
            let kind = expr.kind(operation);
            let label: &'static str = match kind {
                StepKind::Operation(op) => op.into(),
                unexpected => anyhow::bail!("Unexpected step kind: {unexpected:?}"),
            };
            writeln!(w, "  subgraph cluster_{operation:?} {{")?;
            writeln!(w, "  label = \"{operation} = {label}\";")?;
            writeln!(w, "  labelloc = \"B\";")?;
        }

        for id in ids {
            let (kind, children) = expr.node(id);

            let label = node_label(kind);
            writeln!(
                w,
                "  node_{id:?} [label = \"{id:?} = {label}({})\"];",
                children
                    .iter()
                    .format_with(", ", |elt, f| f(&format_args!("{elt:?}")))
            )?;
        }

        if let Some(operation) = operation {
            writeln!(w, "  }} // cluster_{operation:?}")?;
        }

        writeln!(w)?;
    }

    for id in expr.ids() {
        let (_, children) = expr.node(id);

        if !children.is_empty() {
            writeln!(
                w,
                "  {{{}}} -> node_{id:?};",
                children
                    .iter()
                    .format_with(";", |elt, f| f(&format_args!("node_{elt:?}")))
            )?;
        }
    }

    writeln!(w, "}}")?;
    w.flush()?;

    Ok(())
}

fn node_label(kind: &StepKind) -> Cow<'static, str> {
    match kind {
        StepKind::Operation(operation) => Cow::Borrowed(operation.into()),
        StepKind::Expression(Expression::Literal(literal)) => {
            // TODO: Escape quotes in strings
            let label = format!("literal:{literal}");
            Cow::Owned(label)
        }
        StepKind::Expression(Expression::LateBound(late_bound)) => {
            Cow::Borrowed(late_bound.label())
        }
        StepKind::Expression(Expression::Inst(InstKind::Simple(inst))) => {
            Cow::Borrowed(inst.into())
        }
        StepKind::Expression(Expression::Inst(InstKind::FieldRef)) => Cow::Borrowed("field_ref"),
        StepKind::Expression(Expression::Inst(InstKind::Cast(to_type))) => {
            let label = format!("cast:{}", sparrow_syntax::FormatDataType(to_type));
            Cow::Owned(label)
        }
        StepKind::Expression(Expression::Inst(InstKind::Record)) => Cow::Borrowed("record"),
        StepKind::Transform => Cow::Borrowed("transform"),
        StepKind::Error => Cow::Borrowed("error"),
        StepKind::Window(window) => Cow::Borrowed(window.label()),
        StepKind::Expression(Expression::Inst(InstKind::Udf(udf))) => {
            Cow::Owned(udf.signature().name().to_owned())
        }
    }
}
