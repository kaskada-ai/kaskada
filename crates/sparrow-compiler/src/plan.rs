use arrow::datatypes::DataType;
use sparrow_api::kaskada::v1alpha::{ComputePlan, PerEntityBehavior};
use sparrow_core::debug_println;

use crate::dfg::{DfgExpr, Operation, StepKind};
use crate::plan::plan_builder::PlanBuilder;
use crate::DataContext;

mod expression_to_plan;
mod finalize_expression_indices;
mod interpolations;
mod operation_schedule;
mod operation_to_plan;
mod plan_builder;
mod transform_to_plan;

/// A constant to easily enable debug prints int he plan code.
///
/// This will be compiled away when `false`.
const DBG_PRINT_PLAN: bool = false;

/// Extracts a `ComputePlan` proto from a `DfgExpr`.
///
/// TODO: The `DataContext` is used to get the table name from an ID, which is
/// only necessary to create the `slice_plan` because it uses a name instead of
/// an ID.
pub fn extract_plan_proto(
    data_context: &DataContext,
    expr: DfgExpr,
    per_entity_behavior: PerEntityBehavior,
    primary_grouping: String,
    primary_grouping_key_type: &DataType,
) -> anyhow::Result<ComputePlan> {
    // TODO: Projection pushdown?
    // TODO: Slice analysis?

    let mut transform_to_plan = transform_to_plan::TransformToPlan::new(&expr);

    let mut plan_builder = PlanBuilder::try_new(&expr)?;

    for id in expr.ids() {
        let (kind, children) = expr.node(id);

        let operation_index = plan_builder.schedule.operation_opt(id);

        debug_println!(
            DBG_PRINT_PLAN,
            "Adding {id:?}={kind:?}({children:?}) to plan in operation {operation_index:?}",
        );

        let operation_index = if let Some(operation_index) = operation_index {
            operation_index
        } else {
            // We don't add nodes associated with the empty operation to the plan.
            // These should correspond to expressions that weren't dependent on
            // any data, so they should be literals, late values, etc. which are
            // are all added as expressions and transformed into the appropriate
            // domain.
            debug_println!(
                DBG_PRINT_PLAN,
                "Skipping node {id:?}={kind:?} outside operation"
            );
            plan_builder.add_invalid(id, "Node in the empty operation");
            continue;
        };

        match kind {
            StepKind::Operation(Operation::Empty) => {
                anyhow::bail!(
                    "Empty operation should have been skipped earlier, but was in operation index \
                     {operation_index}"
                )
            }
            StepKind::Operation(operation) => {
                let operation = operation_to_plan::dfg_to_plan(
                    &transform_to_plan,
                    data_context,
                    &mut plan_builder,
                    id,
                    operation,
                    children,
                    &expr,
                )?;
                // Add the generated operator to the transform infor.
                transform_to_plan.register_operation(operation_index, operation.operator()?)?;
            }
            StepKind::Expression(expression) => {
                // The last child is the operation. Everything else is an argument.
                let arguments = &children[0..children.len() - 1];

                let expression_index = expression_to_plan::dfg_to_plan(
                    &mut plan_builder,
                    id,
                    operation_index,
                    expression,
                    arguments,
                )?;
                debug_println!(
                    DBG_PRINT_PLAN,
                    "... Added as {operation_index}.{expression_index}"
                );
            }
            StepKind::Transform => {
                let expression_index = transform_to_plan.dfg_to_plan(
                    &mut plan_builder,
                    id,
                    operation_index,
                    children[0],
                )?;
                debug_println!(
                    DBG_PRINT_PLAN,
                    "... Added as {operation_index}.{expression_index}"
                );
            }
            StepKind::Error => {
                anyhow::bail!("Encountered error in DFG, unable to create plan")
            }
            StepKind::Window(_) => todo!(),
        };
    }

    let output_id = egg::Id::from(expr.len() - 1);
    let primary_grouping_key_type = primary_grouping_key_type.try_into()?;
    plan_builder.finish(
        per_entity_behavior,
        output_id,
        primary_grouping,
        primary_grouping_key_type,
    )
}

#[cfg(test)]
mod tests {
    use sparrow_api::kaskada::v1alpha::compile_request::ExpressionKind;
    use sparrow_api::kaskada::v1alpha::SliceRequest;
    use sparrow_api::kaskada::v1alpha::{
        CompileRequest, ComputePlan, FeatureSet, PerEntityBehavior,
    };

    use crate::{DataContext, InternalCompileOptions};

    #[tokio::test]
    async fn test_simple_plan() {
        insta::assert_yaml_snapshot!(create_plan(
            "simple_plan",
            &DataContext::for_test(),
            FeatureSet::new("{ x: Table1.x_i64 + 10 }", vec![]),
            None,
            PerEntityBehavior::All,
        )
        .await
        .unwrap());
    }

    #[tokio::test]
    async fn test_two_tables() {
        insta::assert_yaml_snapshot!(create_plan(
            "two_tables",
            &DataContext::for_test(),
            FeatureSet::new("{ x: Table1.x_i64 + Table2.x_i64 }", vec![]),
            None,
            PerEntityBehavior::All,
        )
        .await
        .unwrap());
    }

    #[tokio::test]
    async fn test_three_tables() {
        insta::assert_yaml_snapshot!(create_plan(
            "three_tables",
            &DataContext::for_test(),
            FeatureSet::new("{ x: Table1.x_i64 + Table2.x_i64 + Table3.x_i64 }", vec![]),
            None,
            PerEntityBehavior::All,
        )
        .await
        .unwrap());
    }

    #[tokio::test]
    async fn test_three_tables_merged_differently() {
        // TODO: Simplify redundant merges
        //
        // This plan currently contains unnecessary merges because it doesn't re-order
        // the the operations. Specifically, the simplifier doesn't know that we
        // can just merge everything to get the merged set of Table1, Table2,
        // and Table3, and then do all the math in that merged domain. This is
        // probably OK for now (doesn't affect correctness, only wasted work).
        insta::assert_yaml_snapshot!(create_plan(
            "three_tables_merged_differently",
            &DataContext::for_test(),
            FeatureSet::new(
                "{ x: (Table1.x_i64 + Table2.x_i64) + Table3.x_i64, y: Table1.x_i64 + \
                 (Table2.x_i64 + Table3.x_i64) }",
                vec![]
            ),
            None,
            PerEntityBehavior::All,
        )
        .await
        .unwrap());
    }

    async fn create_plan(
        name: &str,
        data_context: &DataContext,
        feature_set: FeatureSet,
        slice_request: Option<SliceRequest>,
        per_entity_behavior: PerEntityBehavior,
    ) -> anyhow::Result<ComputePlan> {
        let mut test_output_dir = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        test_output_dir.push("tests");
        test_output_dir.push("test-output");
        test_output_dir.push("plan");

        if !test_output_dir.exists() {
            std::fs::create_dir_all(&test_output_dir).unwrap()
        }

        let response = crate::compile_proto(
            CompileRequest {
                tables: data_context.proto_tables()?,
                feature_set: Some(feature_set),
                slice_request,
                expression_kind: ExpressionKind::Complete as i32,
                experimental: false,
                per_entity_behavior: per_entity_behavior as i32,
            },
            InternalCompileOptions {
                store_final_dfg: Some(test_output_dir.join(format!("{name}_final_dfg.dot"))),
                store_plan_graph: Some(test_output_dir.join(format!("{name}_plan.dot"))),
                ..InternalCompileOptions::default()
            },
        )
        .await
        .map_err(|e| e.into_error())?;

        if let Some(plan) = response.plan {
            Ok(plan)
        } else {
            Err(anyhow::anyhow!(
                "No plan produced. Diagnostics {:?}",
                response.fenl_diagnostics
            ))
        }
    }
}
