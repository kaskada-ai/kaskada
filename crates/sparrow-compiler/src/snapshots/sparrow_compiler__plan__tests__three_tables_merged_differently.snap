---
source: crates/sparrow-compiler/src/plan.rs
expression: "create_plan(\"three_tables_merged_differently\", &DataContext::for_test(),\n            FeatureSet::new(\"{ x: (Table1.x_i64 + Table2.x_i64) + Table3.x_i64, y: Table1.x_i64 + \\\n                 (Table2.x_i64 + Table3.x_i64) }\",\n                vec![]), None, PerEntityBehavior::All).await.unwrap()"
---
per_entity_behavior: 1
operations:
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: x_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: y_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: s_str
                  data_type:
                    kind:
                      Primitive: 14
                  nullable: true
                - name: z_i32
                  data_type:
                    kind:
                      Primitive: 5
                  nullable: true
                - name: f_f64
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: a_bool
                  data_type:
                    kind:
                      Primitive: 2
                  nullable: true
                - name: b_bool
                  data_type:
                    kind:
                      Primitive: 2
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 0
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: x_i64
      - arguments:
          - 0
          - 1
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: field_ref
    operator:
      Scan:
        table_id:
          high: 0
          low: 0
        schema:
          fields:
            - name: x_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
            - name: y_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
            - name: s_str
              data_type:
                kind:
                  Primitive: 14
              nullable: true
            - name: z_i32
              data_type:
                kind:
                  Primitive: 5
              nullable: true
            - name: f_f64
              data_type:
                kind:
                  Primitive: 13
              nullable: true
            - name: a_bool
              data_type:
                kind:
                  Primitive: 2
              nullable: true
            - name: b_bool
              data_type:
                kind:
                  Primitive: 2
              nullable: true
        slice_plan:
          table_name: Table1
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: x_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: y_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 1
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: x_i64
      - arguments:
          - 0
          - 1
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: field_ref
    operator:
      Scan:
        table_id:
          high: 0
          low: 1
        schema:
          fields:
            - name: x_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
            - name: y_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
        slice_plan:
          table_name: Table2
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 0
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 1
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments:
          - 1
          - 0
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: add
    operator:
      Merge:
        left: 1
        right: 0
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: x_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: y_i64
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 3
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: x_i64
      - arguments:
          - 0
          - 1
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: field_ref
    operator:
      Scan:
        table_id:
          high: 0
          low: 2
        schema:
          fields:
            - name: x_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
            - name: y_i64
              data_type:
                kind:
                  Primitive: 6
              nullable: true
        slice_plan:
          table_name: Table3
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 2
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 3
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments:
          - 1
          - 0
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: add
    operator:
      Merge:
        left: 3
        right: 2
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 1
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 3
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments:
          - 1
          - 0
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: add
    operator:
      Merge:
        left: 3
        right: 1
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 0
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 5
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments:
          - 1
          - 0
        result_type:
          kind:
            Primitive: 6
        output: true
        operator:
          Instruction: add
    operator:
      Merge:
        left: 5
        right: 0
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          LateBound: 1
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 6
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: y
      - arguments: []
        result_type:
          kind:
            Primitive: 6
        output: false
        operator:
          Input:
            producing_operation: 4
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 2
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: x
      - arguments:
          - 4
          - 3
          - 2
          - 1
        result_type:
          kind:
            Struct:
              fields:
                - name: x
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: y
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
        output: true
        operator:
          Instruction: record
      - arguments:
          - 5
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          Instruction: time_of
      - arguments:
          - 6
          - 0
        result_type:
          kind:
            Primitive: 2
        output: true
        operator:
          Instruction: gte
    operator:
      Merge:
        left: 6
        right: 4
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: x
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
                - name: y
                  data_type:
                    kind:
                      Primitive: 6
                  nullable: true
        output: true
        operator:
          Input:
            producing_operation: 7
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 5
    operator:
      Select:
        input: 7
        condition:
          producing_operation: 7
          input_column: 4
          interpolation: 1
          column:
            ProducerExpression: 7
primary_grouping: group
primary_grouping_key_type:
  kind:
    Primitive: 6

